### PR TITLE
♻️ project CRUD routes: updates OAS and model schemas (part 1)

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

This PR reviews the specs and implementation of CRUD operations on the ``project`` resource.  The OAS now exposes a PATCH as well and all the schemas have been reviewed:

image

This PR is very verbose and therefore, for the sake of clarity, is the first of many parts.

### Highlights

-  ✨ Reviews **schemas** used  i/o operations with ``project`` and **creates multiple pydantic models for projects in ``projects/_rest_schemas.py``**
   - see ``test_projects__rest_schemas.py`` where new models are tested against data captured from real fe-be communications.
   - OAS auto-generated with ``api/specs/webserver/scripts/openapi_projects_crud.py``
- ✨ New ``PATCH /project/*`` interface (Implementation will follow in next PR)
- ♻️ refactors ``project_crud_handlers``
    -  separated implementation under ``projects/_utils*.py``
    - ``POST /projects`` validate now request body with new models
- 🗑️ ``api/schemas`` files
    -  validation is directly done using pydantic models
- 📝 docs raised expections
-   Upgrade of ``mypy`` in https://github.com/ITISFoundation/osparc-simcore/pull/4108 resurfaced new issues
    - 🔨 fixes sqlalchemy plugin config
    - ♻️ fixes mypy issues in packages/postgres-database
    - ♻️ fixes mypy issues in catalog  

## Related issue/s

- related to #4059
- related to #2980

## How to test

```cmd
cd services/web/server
make install-dev
pytest -vv tests/unit/isolated/test_project*.py
```

### CODE REVIEW

File: packages/models-library/src/models_library/projects.py [hunk 1]
```diff
@@ -17,13 +18,24 @@
 from .projects_state import ProjectState
 from .projects_ui import StudyUI
 
-ProjectID = UUID
-ProjectIDStr = constr(regex=UUID_RE_BASE)
+ProjectID: TypeAlias = UUID
+ClassifierID: TypeAlias = str
 
-ClassifierID = str
+NodesDict: TypeAlias = dict[NodeIDStr, Node]
```
Comment (sanderegg) [inline packages/models-library/src/models_library/projects.py:24]: Why this name change? this is still the Workbench right? can we keep it so? or is there a valid reason for changing the name?

File: packages/models-library/src/models_library/projects.py [hunk 2]
```diff
@@ -17,13 +18,24 @@
...
+
+class ProjectIDStr(ConstrainedStr):
+    regex = re.compile(UUID_RE_BASE)
+
+    class Config:
+        frozen = True
+
+
+class DateTimeStr(ConstrainedStr):
+    regex = re.compile(DATE_RE)
```
Comment (sanderegg) [inline packages/models-library/src/models_library/projects.py:35]: what do we use here? we should go with some ISO if possible.

File: packages/models-library/src/models_library/projects.py [hunk 3]
```diff
@@ -110,17 +122,15 @@ class Project(BaseProjectModel):
...
     )
 
-    # Timestamps   TODO: should we use datetime??
-    creation_date: str = Field(
+    # Timestamps
+    creation_date: DateTimeStr = Field(
         ...,
-        regex=DATE_RE,
         description="project creation date",
         examples=["2018-07-01T11:13:43Z"],
```
Comment (sanderegg) [inline packages/models-library/src/models_library/projects.py:129]: this very much looks like the ISO timestamp.

File: packages/models-library/src/models_library/projects_nodes.py [hunk 4]
```diff
@@ -55,10 +56,20 @@
...
-InputID = OutputID = constr(regex=PROPERTY_KEY_RE)
-InputsDict = dict[InputID, InputTypes]
-OutputsDict = dict[OutputID, OutputTypes]
-UnitStr = constr(strip_whitespace=True)
+
+class KeyIDStr(ConstrainedStr):
+    regex = re.compile(PROPERTY_KEY_RE)
+
+
+InputID: TypeAlias = KeyIDStr
```
Comment (sanderegg) [inline packages/models-library/src/models_library/projects_nodes.py:64]: Question: adding the ```TypeAlias``` type here, does it bring something? just wondering.

File: packages/models-library/src/models_library/projects_nodes.py [hunk 5]
```diff
@@ -55,10 +56,20 @@
...
+OutputID: TypeAlias = KeyIDStr
+
+InputsDict: TypeAlias = dict[InputID, InputTypes]
+OutputsDict: TypeAlias = dict[OutputID, OutputTypes]
+
+
+class UnitStr(ConstrainedStr):
```
Comment (sanderegg) [inline packages/models-library/src/models_library/projects_nodes.py:71]: thanks!

File: packages/service-library/src/servicelib/aiohttp/requests_validation.py [hunk 6]
```diff
@@ -115,7 +137,15 @@ def parse_request_query_parameters_as(
 ) -> ModelType:
     """Parses query parameters from 'request' and validates against 'parameters_schema'
 
-    :raises HTTPUnprocessableEntity (422) if validation of queries fail
+
+    Keyword Arguments:
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/aiohttp/requests_validation.py:141]: Q: do you use some extension to create these docstrings?

**Review by sanderegg — APPROVED** [review]
14k files... you're doing a competition? ;)
looking forward to seeing my PRs in conflicts...

File: packages/service-library/src/servicelib/aiohttp/requests_validation.py (hunk 6 above)
Comment (pcrespov) [reply packages/service-library/src/servicelib/aiohttp/requests_validation.py:141]: yes. it is in the https://github.com/ITISFoundation/osparc-simcore/blob/c032fb5e133da4786d5384860daa4fecf672ab64/.vscode/extensions.json#L6

And check comments on pep257 in https://github.com/ITISFoundation/osparc-simcore/blob/c032fb5e133da4786d5384860daa4fecf672ab64/docs/coding-conventions.md#python

File: packages/models-library/src/models_library/projects_nodes.py (hunk 4 above)
Comment (pcrespov) [reply packages/models-library/src/models_library/projects_nodes.py:64]: yes. It helps mypy in https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases

File: packages/models-library/src/models_library/projects.py (hunk 3 above)
Comment (pcrespov) [reply packages/models-library/src/models_library/projects.py:129]: @sanderegg it was already like that. I did not dare changing it. But I will try in next PR and see what happens :-)

File: packages/models-library/src/models_library/projects.py (hunk 2 above)
Comment (pcrespov) [reply packages/models-library/src/models_library/projects.py:35]: see response below

File: packages/models-library/src/models_library/projects.py (hunk 1 above)
Comment (pcrespov) [reply packages/models-library/src/models_library/projects.py:24]: Yes, it is still the workbench. Before the entry was
```python

class Project(BaseModel):
    # ...
    workbench: Workbench
    # ...
```

and now is 
```python

class Project(BaseModel):
    # ...
    workbench: NodesDict
    # ...
```

which IMO is much more descriptive

**Review by matusdrobuliak66 — APPROVED** [review]
:+1:

File: api/specs/webserver/scripts/openapi_node_meta.py
Comment (matusdrobuliak66) [inline api/specs/webserver/scripts/openapi_node_meta.py:1]: Just a comment, openapi_node_meta.py & openapi_project.py are helper functions that create node-meta-v0.0.1-pydantic.json & project-v0.0.1-pydantic.json. Even though we remove these two schemas for validation, we still use them to create node-meta-v0.0.1-pydantic-converted-clean.yaml & project-v0.0.1-pydantic-converted-clean.yaml which are referenced in project.yaml & service.yaml which are referenced in openapi_projects-projects.yaml & openapi-node-v0.0.1.yaml files. 

Yes we can delete them, but maybe there might be used in the meantime while we fully refactor and remove all the yaml files.

**Review by matusdrobuliak66 — APPROVED** [review]

**Comment by mrnicegyu11** [comment]
thx, ack

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py [hunk 7]
```diff
@@ -92,11 +89,8 @@ def _is_node_dynamic(node_key: str) -> bool:
     return "/dynamic/" in node_key
 
 
-async def validate_project(app: web.Application, project: dict):
-    project_schema = app[APP_JSONSCHEMA_SPECS_KEY]["projects"]
-    await asyncio.get_event_loop().run_in_executor(
-        None, validate_instance, project, project_schema
-    )
+async def validate_project(project: ProjectDict):
+    await asyncio.get_event_loop().run_in_executor(None, parse_obj_as, Project, project)
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/projects/projects_api.py:93]: Why is this async. I don't get it? At this point you can drop all the async part. and make the .
Please also adjust usage

```suggestion
def validate_project(project: ProjectDict) -> None:
    parse_obj_as(Project, project)
```

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py [hunk 8]
```diff
@@ -132,7 +126,7 @@ async def get_project_for_user(
             user_id, project, project_type is ProjectType.TEMPLATE, app
         )
 
-    await validate_project(app, project)
+    await validate_project(project)
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/projects/projects_api.py:129]: which should be here

**Review by GitHK — COMMENTED** [review]
👍 Nice and thanks for the intro to this

**Review by GitHK — APPROVED** [review]

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py (hunk 7 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/projects/projects_api.py:93]: I tried to keep the same function, but I get it is not necessary

Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/projects/projects_api.py:93]: same here

**Pull request merged**